packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -1915,168 +2178,491 @@ class TimePickerDialog extends StatefulWidget {
+        }
+      case TimePickerEntryMode.input:
+      case TimePickerEntryMode.inputOnly:
+        return _kTimePickerMinInputSize;
+    }
+  }
 
-  late final VoidCallback _entryModeListener;
+  Size _dialogSize(BuildContext context, MediaQueryData mediaQuery, ThemeData theme) {
+    final Orientation orientation = _orientation.value ?? mediaQuery.orientation;
+    // Constrain the textScaleFactor to prevent layout issues. Since only some
+    // parts of the time picker scale up with textScaleFactor, we cap the factor
+    // to 1.1 as that provides enough space to reasonably fit all the content.
+    final double textScaleFactor = math.min(mediaQuery.textScaleFactor, 1.1);
 
-  @override
-  void didChangeDependencies() {
-    super.didChangeDependencies();
-    localizations = MaterialLocalizations.of(context);
-    _announceInitialTimeOnce();
-    _announceModeOnce();
+    final Size timePickerSize;
+    switch (_entryMode.value) {
+      case TimePickerEntryMode.dial:
+      case TimePickerEntryMode.dialOnly:
+        switch (orientation) {
+          case Orientation.portrait:
+            timePickerSize = theme.materialTapTargetSize == MaterialTapTargetSize.padded
+                ? _kTimePickerPortraitSize
+                : _kTimePickerPortraitShrinkWrapSize;
+            break;
+          case Orientation.landscape:
+            timePickerSize = Size(
+                _kTimePickerLandscapeSize.width * textScaleFactor,
+                theme.materialTapTargetSize == MaterialTapTargetSize.padded
+                    ? (theme.useMaterial3 ? _kTimePickerLandscapeSize.height : _kTimePickerLandscapeSizeM2.height)
+                    : (theme.useMaterial3
+                        ? _kTimePickerLandscapeShrinkWrapSize.height
+                        : _kTimePickerLandscapeShrinkWrapSizeM2.height));
+            break;
+        }
+        break;
+      case TimePickerEntryMode.input:
+      case TimePickerEntryMode.inputOnly:
+        final _TimePickerDefaults defaultTheme;
+        final bool alwaysUse24HourFormat = MediaQuery.of(context).alwaysUse24HourFormat;
+        if (theme.useMaterial3) {
+          defaultTheme = _TimePickerDefaultsM3(context);
+          double timePickerWidth;
+          if (alwaysUse24HourFormat) {
+            timePickerWidth = _kTimePickerInputSize.width - defaultTheme.dayPeriodPortraitSize.width - 12;
+          } else {
+            timePickerWidth = _kTimePickerInputSize.width;
+          }
+          timePickerSize = Size(timePickerWidth, _kTimePickerInputSize.height);
+        } else {
+          defaultTheme = _TimePickerDefaultsM2(context);
+          double timePickerWidth;
+          if (alwaysUse24HourFormat) {
+            timePickerWidth = _kTimePickerInputSize.width - defaultTheme.dayPeriodPortraitSize.width - 12;
+          } else {
+            timePickerWidth = _kTimePickerInputSize.width;
+          }
+          timePickerSize = Size(timePickerWidth, _kTimePickerInputSize.height);
+        }
```

Comment:
This looks duplicated. The only difference between M2 and M3 is just the defaults, right? 

Can't we just use the right default theme (M2/M3) and use the same block of code for both M3 and M2? Or am I missing something that is different in the two cases?

Reply:
Good catch, I didn't notice that!  I consolidated it.